docs/TESTPLAN.md

```diff
@@ -0,0 +1,204 @@
+
+Currently, CloudFormationUtils, SdkUtils, and BeanstalkUtils have unit tests testing success
+and failure paths for every function contained in them.
+
+As they exist today, the current Unit Tests do not have any obvious gaps. Every part of the project that is unit testable has tests
+that are meant to succeed and fail, and they have high (>80%) branch coverage.
+
+#### Testing Gaps
+
+The biggest gap in unit testing is that more modules need to be refactored so unit tests can be added: only a small
+percentage of the project has any unit tests.
+
+### Functional Tests
+
+Functional tests make up the majority of tests in the project. They test how a task and all of its components come together
+to work. In general, the point of functional tests is to make sure all of the units work together, and in this project they
+make sure that each Task is interacting properly with the service clients and settings. This is accomplished by passing in
+carefully crafted settings, and mocked implementations of service clients.
+
+For small tasks, the functional tests have great branch coverage without being too complicated, but for larger tasks like
+Beanstalk and CloudFormation, The tests are much longer. Therefore, it is preferable to grow the number of unit
+tests over the number of functional tests as they currently serve the purpose of both making sure units work and function well
+together.
+
+Looking back at past releases that had issues, the second most common reason for issues was bad handling of parameters. The
+functional tests are great at catching this category of problem.
+
+All of the functional tests follow the pattern:
+
+1. Create a `TaskParameters` object for the task
+2. Modify `TaskParameters` to make it go through the desired code path
+3. Create mocked service clients
+4. Run `TaskName.execute()`
+5. Verify service calls/failures on tasks that are supposed to fail
+
+#### Testing Gaps
+
+The following modules do not have any functional tests:
+
+-   CloudFormation Create Or Update
+-   Beanstalk Create Application
+
+The following modules do not have functional tests, but it would make no sense to add them:
+
+-   AWS Powershell
+-   AWS CLI
+
+The other major gap is most functional tests only have ~50% branch coverage. This can be improved by improving our unit tests
+instead.
+
+### End to End Tests
+
+End to end tests serve the purpose of making sure the tasks work when actually connecting to AWS. The goal of them is not to
+test every code path, but to make sure we can:
+
+-   Connect to AWS at all/create clients that work
+-   Instance and fixed credentials both work
+-   Any task that does heavy build/filesystem work is tested more heavily here (Lambda, Cloudformation, and beanstalk tasks)
+-   Anything that relies on the real environment is tested (like the aws shell script task or the powershell scripts)
+
+With this in mind, the goals for functional tests are: to run every task (to see if they work at all), test instance
```

Comment:
* " the goals for functional tests " or " the goals for end to end tests " ?
* "and to test all of as many code paths of tasks like aws shell script as possible." -> "and to test as many code paths within tasks as possible, particularly those that blah blah blah, like aws schell script."
* consider putting each goal as a list item, because the next statement references "the first goal"

Reply:
changed all of those things